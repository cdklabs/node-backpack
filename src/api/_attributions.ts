import * as path from 'path';
import * as fs from 'fs-extra';
import type { ModuleInfo } from 'license-checker';
import { shell } from './_shell';
import type { Package } from './bundle';
import { Violation, ViolationType, ViolationsReport } from './violation';


const ATTRIBUTION_SEPARATOR = '\n----------------\n';

/**
 * Properties for `Attributions`.
 */
export interface AttributionsProps {
  /**
   * The package root directory.
   */
  readonly packageDir: string;
  /**
   * The name of the package.
   */
  readonly packageName: string;
  /**
   * Package dependencies.
   */
  readonly dependencies: Package[];
  /**
   * The parent directory underwhich all dependencies live.
   */
  readonly dependenciesRoot: string;
  /**
   * Path to the attribution licenses file to be created / validated.
   */
  readonly licensesPath: string;
  /**
   * List of allowed licenses.
   *
   */
  readonly allowedLicenses: string[];
  /**
   * Dependencies matching this pattern will be excluded from attribution.
   *
   * @default - no exclusions.
   */
  readonly exclude?: string;
  /**
   * Path to the attribution versions file to be created / validated.
   * If this property is set, dependency versions are left out of the
   * licenses document, and outputed into this file.
   *
   * @default - versions are encoded inside the licenses file.
   */
  readonly versionsPath?: string;
}

/**
 * `Attributions` represents attributions files containing third-party license information.
 */
export class Attributions {

  private readonly packageDir: string;
  private readonly packageName: string;
  private readonly dependencies: Package[];
  private readonly allowedLicenses: string[];
  private readonly dependenciesRoot: string;
  private readonly licensesPath: string;
  private readonly versionsPath?: string;

  private readonly attributions: Attribution[];
  private readonly licenses: string;
  private readonly versions: string;

  constructor(props: AttributionsProps) {
    this.packageDir = props.packageDir;
    this.packageName = props.packageName;
    this.licensesPath = path.join(this.packageDir, props.licensesPath);
    this.versionsPath = props.versionsPath ? path.join(this.packageDir, props.versionsPath) : undefined;
    this.dependencies = props.dependencies.filter(d => !props.exclude || !new RegExp(props.exclude).test(d.name));
    this.allowedLicenses = props.allowedLicenses.map(l => l.toLowerCase());
    this.dependenciesRoot = props.dependenciesRoot;

    // without the generated notice content, this object is pretty much
    // useless, so lets generate those of the bat.
    this.attributions = this.attribute();
    const { licenses, versions } = this.render(this.attributions);

    this.licenses = licenses;
    this.versions = versions;
  }

  /**
   * Validate the current notice file.
   *
   * This method never throws. The Caller is responsible for inspecting the report returned and act accordinagly.
   */
  public validate(): ViolationsReport {

    const violations: Violation[] = [];
    const relLicensesPath = path.relative(this.packageDir, this.licensesPath);
    const relVersionsPath = this.versionsPath ? path.relative(this.packageDir, this.versionsPath) : undefined;

    const fix = () => this.flush();

    const missingLicenses = !fs.existsSync(this.licensesPath);
    const missingVersions = this.versionsPath && !fs.existsSync(this.versionsPath);
    const licenses = missingLicenses ? undefined : fs.readFileSync(this.licensesPath, { encoding: 'utf-8' });
    const versions = (missingVersions || !this.versionsPath) ? undefined : fs.readFileSync(this.versionsPath, { encoding: 'utf-8' });
    const outdatedLicenses = licenses !== undefined && licenses !== this.licenses;
    const outdatedVersions = versions !== undefined && versions !== this.versions;

    if (missingLicenses) {
      violations.push({ type: ViolationType.MISSING_LICENSES, message: `${relLicensesPath} is missing`, fix });
    }

    if (outdatedLicenses) {
      violations.push({ type: ViolationType.OUTDATED_LICENSES, message: `${relLicensesPath} is outdated`, fix });
    }

    if (missingVersions) {
      violations.push({ type: ViolationType.MISSING_VERSIONS, message: `${relVersionsPath} is missing`, fix });
    }

    if (outdatedVersions) {
      violations.push({ type: ViolationType.OUTDATED_VERSIONS, message: `${relVersionsPath} is outdated`, fix });
    }

    const invalidLicense: Violation[] = Array.from(this.attributions.values())
      .filter(a => a.licenses.length === 1 && !this.allowedLicenses.includes(a.licenses[0].toLowerCase()))
      .map(a => ({ type: ViolationType.INVALID_LICENSE, message: `Dependency ${a.packageFqn} has an invalid license: ${a.licenses[0]}` }));

    const noLicense: Violation[] = Array.from(this.attributions.values())
      .filter(a => a.licenses.length === 0)
      .map(a => ({ type: ViolationType.NO_LICENSE, message: `Dependency ${a.packageFqn} has no license` }));

    const multiLicense: Violation[] = Array.from(this.attributions.values())
      .filter(a => a.licenses.length > 1)
      .map(a => ({ type: ViolationType.MULTIPLE_LICENSE, message: `Dependency ${a.packageFqn} has multiple licenses: ${a.licenses}` }));

    violations.push(...invalidLicense);
    violations.push(...noLicense);
    violations.push(...multiLicense);

    return new ViolationsReport(violations);
  }

  /**
   * Flush the generated attributions files to disk.
   */
  public flush() {
    fs.writeFileSync(this.licensesPath, this.licenses);
    if (this.versionsPath) {
      fs.writeFileSync(this.versionsPath, this.versions);
    }
  }

  private render(attributions: Attribution[]): { licenses: string; versions: string } {

    const content = [];

    if (attributions.length > 0) {
      content.push(`The ${this.packageName} package includes the following third-party software/licensing:`);
      content.push('');
    }

    // sort the attributions so the file doesn't change due to ordering issues
    const ordered = Array.from(attributions.values()).sort((a1, a2) => a1.packageFqn.localeCompare(a2.packageFqn));

    const versions: { [key: string]: string[] } = {};

    for (const attr of ordered) {
      const title = this.versionsPath ? attr.packageName : attr.packageFqn;
      content.push(`** ${title} - ${attr.url} | ${attr.licenses[0]}`);

      const versionsInUse = versions[attr.packageName] ?? [];
      versionsInUse.push(attr.packageVersion);
      versions[attr.packageName] = versionsInUse;

      // prefer notice over license
      if (attr.noticeText) {
        content.push(attr.noticeText);
      } else if (attr.licenseText) {
        content.push(attr.licenseText);
      }
      content.push(ATTRIBUTION_SEPARATOR);
    }

    return {
      licenses: content
      // since we are embedding external files, those can different line
      // endings, so we standardize to LF.
        .map(l => l.replace(/\r\n/g, '\n'))
        .join('\n'),
      versions: JSON.stringify(versions, null, 2),
    };

  }

  private attribute(): Attribution[] {

    if (this.dependencies.length === 0) {
      return [];
    }

    const attributions: Attribution[] = [];

    const pkg = (d: Package) => `${d.name}@${d.version}`;

    const packages = this.dependencies.map(d => pkg(d));

    function fetchInfos(_cwd: string, _packages: string[]) {
      // we don't use the programmatic API since it only offers an async API.
      // prefer to stay sync for now since its easier to integrate with other tooling.
      // will offer an async API further down the road.
      const command = `${require.resolve('license-checker/bin/license-checker')} --json --packages "${_packages.join(';')}"`;
      const output = shell(command, { cwd: _cwd, quiet: true });
      return JSON.parse(output);
    }

    // first run a global command to fetch as much information in one shot
    const infos = fetchInfos(this.dependenciesRoot, packages);

    for (const dep of this.dependencies) {
      const key = pkg(dep);

      // sometimes the dependency might not exist from fetching information globally,
      // so we try fetching a concrete package. this can happen for example when
      // two different major versions exist of the same dependency.
      const info: ModuleInfo = infos[key] ?? fetchInfos(dep.path, [pkg(dep)])[key];

      if (!info) {
        // make sure all dependencies are accounted for.
        throw new Error(`Unable to locate license information for ${key} (${dep.path})`);
      }

      const noticeText = info.noticeFile ? fs.readFileSync(info.noticeFile, { encoding: 'utf-8' }) : undefined;

      // for some reason, the license-checker package falls back to the README.md file of the package for license
      // text. this seems strange, disabling that for now.
      // see https://github.com/davglass/license-checker/blob/master/lib/license-files.js#L9
      // note that a non existing license file is ok as long as the license type could be extracted.
      const licenseFile = info.licenseFile?.toLowerCase().endsWith('.md') ? undefined : info.licenseFile;
      const licenseText = licenseFile ? fs.readFileSync(licenseFile, { encoding: 'utf-8' }) : undefined;

      // the licenses key comes in different types but we convert it here
      // to always be an array.
      const licenses = !info.licenses ? undefined : (Array.isArray(info.licenses) ? info.licenses : [info.licenses]);

      const baseUrl = `https://www.npmjs.com/package/${dep.name}`;
      const url = this.versionsPath ? baseUrl : `${baseUrl}/v/${dep.version}`;

      attributions.push({
        packageFqn: key,
        packageName: dep.name,
        packageVersion: dep.version,
        url,
        licenses: licenses ?? [],
        licenseText,
        noticeText,
      });
    }

    return attributions;
  }

}

/**
 * Attribution of a specific dependency.
 */
interface Attribution {
  /**
   * Attributed package fqn (name + version).
   */
  readonly packageFqn: string;
  /**
   * Attributed package name.
   */
  readonly packageName: string;
  /**
   * Attributed package version.
   */
  readonly packageVersion: string;
  /**
   * URL to the package.
   */
  readonly url: string;
  /**
   * Package licenses.
   *
   * Note that some packages will may have multiple licenses,
   * which is why this is an array. In such cases, the license
   * validation will fail since we currently disallow this.
   */
  readonly licenses: string[];
  /**
   * Package license content.
   *
   * In case a package has multiple licenses, this will
   * contain...one of them. It currently doesn't matter which
   * one since it will not pass validation anyway.
   */
  readonly licenseText?: string;
  /**
   * Package notice.
   */
  readonly noticeText?: string;
}
