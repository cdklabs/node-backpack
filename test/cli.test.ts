import * as os from 'os';
import * as path from 'path';
import * as fs from 'fs-extra';
import { shell } from '../src/api/_shell';
import { Package } from './_package';

test('validate', () => {

  const pkg = Package.create({ name: 'consumer', licenses: ['Apache-2.0'], circular: true, notice: 'outdated' });
  const dep1 = pkg.addDependency({ name: 'dep1', licenses: ['INVALID'] });
  const dep2 = pkg.addDependency({ name: 'dep2', licenses: ['Apache-2.0', 'MIT'] });

  pkg.write();
  pkg.install();

  try {
    const command = [
      whereami(),
      '--entrypoint', pkg.entrypoint,
      '--resource', 'missing:bin/missing',
      '--license', 'Apache-2.0',
      'validate',
    ].join(' ');
    shell(command, { cwd: pkg.dir, quiet: true });
  } catch (e: any) {
    const violations = new Set(e.stderr.toString().trim().split('\n').filter((l: string) => l.startsWith('-')));
    const expected = new Set([
      `- invalid-license: Dependency ${dep1.name}@${dep1.version} has an invalid license: UNKNOWN`,
      `- multiple-license: Dependency ${dep2.name}@${dep2.version} has multiple licenses: Apache-2.0,MIT`,
      '- outdated-attributions: THIRD_PARTY_LICENSES is outdated (fixable)',
      '- missing-resource: Unable to find resource (missing) relative to the package directory',
      '- circular-import: lib/bar.js -> lib/foo.js',
    ]);
    expect(violations).toEqual(expected);
  }

});

test('write', () => {

  const pkg = Package.create({ name: 'consumer', licenses: ['Apache-2.0'] });
  pkg.addDependency({ name: 'dep1', licenses: ['MIT'] });
  pkg.addDependency({ name: 'dep2', licenses: ['Apache-2.0'] });

  pkg.write();
  pkg.install();

  const command = [
    whereami(),
    '--entrypoint', pkg.entrypoint,
    '--license', 'Apache-2.0',
    '--license', 'MIT',
    'write',
  ].join(' ');
  const bundleDir = shell(command, { cwd: pkg.dir, quiet: true });

  expect(fs.existsSync(path.join(bundleDir, pkg.entrypoint))).toBeTruthy();
  expect(fs.existsSync(path.join(bundleDir, 'package.json'))).toBeTruthy();
  expect(fs.existsSync(path.join(bundleDir, 'lib', 'foo.js'))).toBeTruthy();
  expect(fs.existsSync(path.join(bundleDir, 'lib', 'bar.js'))).toBeTruthy();
  expect(fs.existsSync(path.join(bundleDir, 'node_modules'))).toBeFalsy();
  expect(fs.existsSync(path.join(bundleDir, '.git'))).toBeFalsy();

  const manifest = fs.readJSONSync(path.join(bundleDir, 'package.json'));
  const entrypoint = fs.readFileSync(path.join(bundleDir, pkg.entrypoint), { encoding: 'utf-8' });

  expect(entrypoint).toMatchSnapshot();
  expect(Object.keys(manifest.devDependencies)).toEqual(['dep1', 'dep2']);
  expect(manifest.dependencies).toEqual({});

});

test('pack', () => {

  const pkg = Package.create({ name: 'consumer', licenses: ['Apache-2.0'] });
  pkg.addDependency({ name: 'dep1', licenses: ['MIT'] });
  pkg.addDependency({ name: 'dep2', licenses: ['Apache-2.0'] });

  pkg.write();
  pkg.install();

  // we need to first fix all violations
  // before we can pack
  const fix = [
    whereami(),
    '--entrypoint', pkg.entrypoint,
    '--license', 'Apache-2.0',
    '--license', 'MIT',
    'validate --fix',
  ].join(' ');
  shell(fix, { cwd: pkg.dir, quiet: true });

  const pack = [
    whereami(),
    '--entrypoint', pkg.entrypoint,
    '--license', 'Apache-2.0',
    '--license', 'MIT',
    'pack',
  ].join(' ');
  shell(pack, { cwd: pkg.dir, quiet: true });

  const tarball = path.join(pkg.dir, `${pkg.name}-${pkg.version}.tgz`);

  const workdir = fs.mkdtempSync(os.tmpdir());
  shell(`npm install ${tarball}`, { cwd: workdir });

  const installed = path.join(workdir, 'node_modules', pkg.name);
  const attributions = fs.readFileSync(path.join(installed, 'THIRD_PARTY_LICENSES'), { encoding: 'utf-8' });
  const versions = fs.readFileSync(path.join(installed, 'THIRD_PARTY_LICENSES.versions.json'), { encoding: 'utf-8' });

  expect(attributions).toMatchSnapshot();
  expect(versions).toMatchSnapshot();

});

test('pack with versions encoded in attributions', () => {

  const pkg = Package.create({ name: 'consumer', licenses: ['Apache-2.0'] });
  pkg.addDependency({ name: 'dep1', licenses: ['MIT'] });
  pkg.addDependency({ name: 'dep2', licenses: ['Apache-2.0'] });

  pkg.write();
  pkg.install();

  // we need to first fix all violations
  // before we can pack
  const fix = [
    whereami(),
    '--entrypoint', pkg.entrypoint,
    '--encode-versions',
    '--license', 'Apache-2.0',
    '--license', 'MIT',
    'validate --fix',
  ].join(' ');
  shell(fix, { cwd: pkg.dir, quiet: true });

  const pack = [
    whereami(),
    '--entrypoint', pkg.entrypoint,
    '--encode-versions',
    '--license', 'Apache-2.0',
    '--license', 'MIT',
    'pack',
  ].join(' ');
  shell(pack, { cwd: pkg.dir, quiet: true });

  const tarball = path.join(pkg.dir, `${pkg.name}-${pkg.version}.tgz`);

  const workdir = fs.mkdtempSync(os.tmpdir());
  shell(`npm install ${tarball}`, { cwd: workdir });

  const installed = path.join(workdir, 'node_modules', pkg.name);
  const attributions = fs.readFileSync(path.join(installed, 'THIRD_PARTY_LICENSES'), { encoding: 'utf-8' });

  expect(fs.existsSync(path.join(installed, 'THIRD_PARTY_LICENSES.versions.json'))).toBeFalsy();
  expect(attributions).toMatchSnapshot();

});

function whereami() {
  return path.join(path.join(__dirname, '..', 'bin', 'node-backpack'));
}