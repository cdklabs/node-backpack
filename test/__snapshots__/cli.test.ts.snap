// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pack 1`] = `
"The consumer package includes the following third-party software/licensing:

** dep1 - https://www.npmjs.com/package/dep1/v/0.0.1 | MIT

----------------

** dep2 - https://www.npmjs.com/package/dep2/v/0.0.1 | Apache-2.0

----------------
"
`;

exports[`pack 2`] = `
"{
  \\"dep1\\": [
    \\"0.0.1\\"
  ],
  \\"dep2\\": [
    \\"0.0.1\\"
  ]
}"
`;

exports[`write 1`] = `
"var __getOwnPropNames = Object.getOwnPropertyNames;
var __commonJS = (cb, mod) => function __require() {
  return mod || (0, cb[__getOwnPropNames(cb)[0]])((mod = { exports: {} }).exports, mod), mod.exports;
};

// lib/foo.js
var require_foo = __commonJS({
  \\"lib/foo.js\\"() {
    console.log(\\"hello from foo\\");
  }
});

// lib/bar.js
var require_bar = __commonJS({
  \\"lib/bar.js\\"() {
    console.log(\\"hello from bar\\");
  }
});

// node_modules/dep1/lib/foo.js
var require_foo2 = __commonJS({
  \\"node_modules/dep1/lib/foo.js\\"() {
    console.log(\\"hello from foo\\");
  }
});

// node_modules/dep1/lib/bar.js
var require_bar2 = __commonJS({
  \\"node_modules/dep1/lib/bar.js\\"() {
    console.log(\\"hello from bar\\");
  }
});

// node_modules/dep1/lib/index.js
var require_lib = __commonJS({
  \\"node_modules/dep1/lib/index.js\\"() {
    require_foo2();
    require_bar2();
  }
});

// node_modules/dep2/lib/foo.js
var require_foo3 = __commonJS({
  \\"node_modules/dep2/lib/foo.js\\"() {
    console.log(\\"hello from foo\\");
  }
});

// node_modules/dep2/lib/bar.js
var require_bar3 = __commonJS({
  \\"node_modules/dep2/lib/bar.js\\"() {
    console.log(\\"hello from bar\\");
  }
});

// node_modules/dep2/lib/index.js
var require_lib2 = __commonJS({
  \\"node_modules/dep2/lib/index.js\\"() {
    require_foo3();
    require_bar3();
  }
});

// lib/index.js
require_foo();
require_bar();
require_lib();
require_lib2();
//# sourceMappingURL=data:application/json;base64,ewogICJ2ZXJzaW9uIjogMywKICAic291cmNlcyI6IFsiZm9vLmpzIiwgImJhci5qcyIsICIuLi9ub2RlX21vZHVsZXMvZGVwMS9saWIvZm9vLmpzIiwgIi4uL25vZGVfbW9kdWxlcy9kZXAxL2xpYi9iYXIuanMiLCAiLi4vbm9kZV9tb2R1bGVzL2RlcDEvbGliL2luZGV4LmpzIiwgIi4uL25vZGVfbW9kdWxlcy9kZXAyL2xpYi9mb28uanMiLCAiLi4vbm9kZV9tb2R1bGVzL2RlcDIvbGliL2Jhci5qcyIsICIuLi9ub2RlX21vZHVsZXMvZGVwMi9saWIvaW5kZXguanMiLCAiaW5kZXguanMiXSwKICAic291cmNlc0NvbnRlbnQiOiBbImNvbnNvbGUubG9nKFwiaGVsbG8gZnJvbSBmb29cIik7IiwgImNvbnNvbGUubG9nKFwiaGVsbG8gZnJvbSBiYXJcIik7IiwgImNvbnNvbGUubG9nKFwiaGVsbG8gZnJvbSBmb29cIik7IiwgImNvbnNvbGUubG9nKFwiaGVsbG8gZnJvbSBiYXJcIik7IiwgInJlcXVpcmUoXCIuL2Zvb1wiKTtcbnJlcXVpcmUoXCIuL2JhclwiKTsiLCAiY29uc29sZS5sb2coXCJoZWxsbyBmcm9tIGZvb1wiKTsiLCAiY29uc29sZS5sb2coXCJoZWxsbyBmcm9tIGJhclwiKTsiLCAicmVxdWlyZShcIi4vZm9vXCIpO1xucmVxdWlyZShcIi4vYmFyXCIpOyIsICJyZXF1aXJlKFwiLi9mb29cIik7XG5yZXF1aXJlKFwiLi9iYXJcIik7XG5yZXF1aXJlKFwiZGVwMVwiKTtcbnJlcXVpcmUoXCJkZXAyXCIpOyJdLAogICJtYXBwaW5ncyI6ICI7Ozs7OztBQUFBO0FBQUE7QUFBQSxZQUFRLElBQUksZ0JBQWdCO0FBQUE7QUFBQTs7O0FDQTVCO0FBQUE7QUFBQSxZQUFRLElBQUksZ0JBQWdCO0FBQUE7QUFBQTs7O0FDQTVCO0FBQUE7QUFBQSxZQUFRLElBQUksZ0JBQWdCO0FBQUE7QUFBQTs7O0FDQTVCO0FBQUE7QUFBQSxZQUFRLElBQUksZ0JBQWdCO0FBQUE7QUFBQTs7O0FDQTVCO0FBQUE7QUFBQTtBQUNBO0FBQUE7QUFBQTs7O0FDREE7QUFBQTtBQUFBLFlBQVEsSUFBSSxnQkFBZ0I7QUFBQTtBQUFBOzs7QUNBNUI7QUFBQTtBQUFBLFlBQVEsSUFBSSxnQkFBZ0I7QUFBQTtBQUFBOzs7QUNBNUI7QUFBQTtBQUFBO0FBQ0E7QUFBQTtBQUFBOzs7QUNEQTtBQUNBO0FBQ0E7QUFDQTsiLAogICJuYW1lcyI6IFtdCn0K
"
`;
